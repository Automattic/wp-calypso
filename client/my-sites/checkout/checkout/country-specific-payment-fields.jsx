/**
 * External dependencies
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { localize } from 'i18n-calypso';
import { find, isEmpty, includes, get } from 'lodash';

/**
 * Internal dependencies
 */
import FormPhoneMediaInput from 'components/forms/form-phone-media-input';
import { StateSelect, Input, HiddenInput } from 'my-sites/domains/components/form';
import { PAYMENT_PROCESSOR_COUNTRIES_FIELDS } from 'lib/checkout/constants';
import InfoPopover from 'components/info-popover';
import { paymentMethodName } from 'lib/cart-values';

export class CountrySpecificPaymentFields extends Component {
	static propTypes = {
		countryCode: PropTypes.string.isRequired,
		countriesList: PropTypes.array.isRequired,
		getErrorMessage: PropTypes.func.isRequired,
		getFieldValue: PropTypes.func.isRequired,
		handleFieldChange: PropTypes.func.isRequired,
		fieldClassName: PropTypes.string,
		disableFields: PropTypes.bool,
	};

	static defaultProps = {
		fieldClassName: '',
		disableFields: false,
	};

	constructor( props ) {
		super( props );
		this.state = {
			userSelectedPhoneCountryCode: '',
			fields: get( PAYMENT_PROCESSOR_COUNTRIES_FIELDS[ props.countryCode ], 'fields', null ),
		};
	}

	componentDidUpdate( prevProps ) {
		if ( prevProps.countryCode !== this.props.countryCode ) {
			this.setFieldsState( this.props.countryCode );
		}
	}

	setFieldsState( countryCode ) {
		this.setState( {
			fields: get( PAYMENT_PROCESSOR_COUNTRIES_FIELDS[ countryCode ], 'fields', null ),
		} );
	}

	createField = ( fieldName, componentClass, props ) => {
		const errorMessage = this.props.getErrorMessage( fieldName ) || [];
		const isError = ! isEmpty( errorMessage );
		const fieldProps = Object.assign(
			{},
			{
				additionalClasses: `checkout__checkout-field ${ this.props.fieldClassName }`,
				isError,
				errorMessage: errorMessage[ 0 ],
				name: fieldName,
				onBlur: this.onFieldChange,
				onChange: this.onFieldChange,
				value: this.props.getFieldValue( fieldName ) || '',
				autoComplete: 'off',
				labelClass: 'checkout__form-label',
				disabled: this.props.disableFields,
			},
			props
		);

		return includes( this.state.fields, fieldName )
			? React.createElement( componentClass, fieldProps )
			: null;
	};

	handlePhoneFieldChange = ( { value, countryCode } ) => {
		this.props.handleFieldChange( 'phone-number', value );
		this.setState( { userSelectedPhoneCountryCode: countryCode } );
	};

	getPanNumberPopover = () => {
		const { translate } = this.props;
		const popoverText = translate( 'Paying with %(indiaPaymentMethods)s requires a PAN number.', {
			comment: 'indiaPaymentMethods are local payment methods in India.',
			args: {
				indiaPaymentMethods: paymentMethodName( 'netbanking' ),
			},
		} );
		return (
			<InfoPopover position="right" className="checkout__pan-number-popover">
				{ popoverText }
			</InfoPopover>
		);
	};

	onFieldChange = event => this.props.handleFieldChange( event.target.name, event.target.value );

	render() {
		const { translate, countriesList, countryCode } = this.props;
		const { userSelectedPhoneCountryCode } = this.state;
		const countryData = find( countriesList, { code: countryCode } );
		const countryName = countryData && countryData.name ? countryData.name : '';
		const containerClassName = classNames(
			'checkout__country-payment-fields',
			`checkout__country-${ countryCode.toLowerCase() }`
		);

		return (
			<div className={ containerClassName }>
				<span key="country-required-fields" className="checkout__form-info-text">
					{ countryName &&
						translate( 'The following fields are also required for payments in %(countryName)s', {
							args: {
								countryName,
							},
						} ) }
				</span>

				{ this.createField( 'document', Input, {
					label: translate( 'Taxpayer Identification Number', {
						comment:
							'Individual taxpayer registry identification required ' +
							'for Brazilian payment methods on credit card form',
					} ),
				} ) }

				{ this.createField( 'nik', Input, {
					label: translate( 'NIK - Indonesia Identity Card Number', {
						comment:
							'NIK - Indonesia Identity Card Number required ' + 'for Indonesian payment methods.',
					} ),
				} ) }

				{ this.createField( 'pan', Input, {
					placeholder: ' ',
					label: translate( 'PAN Number {{panNumberPopover/}}', {
						comment: 'India PAN number ',
						components: {
							panNumberPopover: this.getPanNumberPopover(),
						},
					} ),
				} ) }

				{ this.createField( 'phone-number', FormPhoneMediaInput, {
					onChange: this.handlePhoneFieldChange,
					countriesList,
					// If the user has manually selected a country for the phone
					// number, use that, but otherwise default this to the same
					// country as the billing address.
					countryCode: userSelectedPhoneCountryCode || countryCode,
					label: translate( 'Phone' ),
				} ) }

				{ this.createField( 'address-1', Input, {
					maxLength: 40,
					label: translate( 'Address' ),
				} ) }

				{ this.createField( 'street-number', Input, {
					inputMode: 'numeric',
					label: translate( 'Street Number', {
						comment: 'Street number associated with address on credit card form',
					} ),
				} ) }

				{ this.createField( 'address-2', HiddenInput, {
					maxLength: 40,
					label: translate( 'Address Line 2' ),
					text: translate( '+ Add Address Line 2' ),
				} ) }

				{ this.createField( 'city', Input, {
					label: translate( 'City' ),
				} ) }

				<div className="checkout__form-state-field">
					{ this.createField( 'state', StateSelect, {
						countryCode,
						label: translate( 'State' ),
					} ) }
				</div>

				{ this.createField( 'postal-code', Input, {
					label: translate( 'Postal Code' ),
				} ) }
			</div>
		);
	}
}

export default localize( CountrySpecificPaymentFields );
