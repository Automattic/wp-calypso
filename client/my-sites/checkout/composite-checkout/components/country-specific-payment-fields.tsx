import styled from '@emotion/styled';
import classNames from 'classnames';
import { useTranslate } from 'i18n-calypso';
import { createElement, useState } from 'react';
import FormPhoneMediaInput from 'calypso/components/forms/form-phone-media-input';
import InfoPopover from 'calypso/components/info-popover';
import { paymentMethodName } from 'calypso/lib/cart-values';
import { PAYMENT_PROCESSOR_COUNTRIES_FIELDS } from 'calypso/lib/checkout/constants';
import { StateSelect, Input, HiddenInput } from 'calypso/my-sites/domains/components/form';
import type { CountryListItem } from '@automattic/wpcom-checkout';
import type { ComponentType } from 'react';

export type CountrySpecificPaymentFieldsProps = {
	countriesList: CountryListItem[];
	countryCode: string;
	disableFields?: boolean;
	getErrorMessages: ( fieldName: string ) => string[];
	getFieldValue: ( fieldName: string ) => string;
	handleFieldChange: ( fieldName: string, newValue: string ) => void;
};

export const CountrySpecificPaymentFields = styled( CountrySpecificPaymentFieldsUnstyled )`
	margin-top: 0;

	& .checkout__form-info-text {
		display: none;
	}

	& .document,
	& .phone,
	& .postal-code,
	& .checkout__form-state-field {
		flex-basis: auto;
	}

	& .checkout__checkout-field,
	& .checkout__form-state-field {
		margin-left: 0;
	}

	& .checkout__form-info-text,
	& .form__hidden-input a {
		margin-left: 0;
	}

	@media ( ${ ( props ) => props.theme.breakpoints.bigPhoneUp } ) {
		& .address-1,
		& .state {
			margin-right: 15px;
		}
	}
`;

function getFieldNamesForCountry( countryCode: string ): string[] {
	if ( countryCode === 'IN' || countryCode === 'BR' ) {
		return PAYMENT_PROCESSOR_COUNTRIES_FIELDS[ countryCode ]?.fields ?? [];
	}
	return [];
}

interface CreateFieldArgs< P > {
	fieldName: string;
	componentClass: ComponentType< P >;
	props: P;
	value: string;
	errorMessage: string;
	handleFieldChange: ( fieldName: string, newValue: string ) => void;
	disabled: boolean;
}

function createField< P >( {
	fieldName,
	componentClass,
	props,
	value,
	errorMessage,
	handleFieldChange,
	disabled,
}: CreateFieldArgs< P > ) {
	const onFieldChange = ( event: { target: { name: string; value: string } } ) =>
		handleFieldChange( event.target.name, event.target.value );
	const fieldProps = Object.assign(
		{},
		{
			additionalClasses: 'checkout__checkout-field',
			isError: errorMessage.length > 0,
			errorMessage,
			name: fieldName,
			onBlur: onFieldChange,
			onChange: onFieldChange,
			value,
			autoComplete: 'off',
			labelClass: 'checkout__form-label',
			disabled,
		},
		props
	);

	return createElement( componentClass, fieldProps );
}

interface CountrySpecificPaymentFieldProps< P > {
	fieldName: string;
	allowedFieldNames: string[];
	componentClass: ComponentType< P >;
	props: P;
	disabled?: boolean;
	getErrorMessages: ( fieldName: string ) => string[];
	getFieldValue: ( fieldName: string ) => string;
	handleFieldChange: ( fieldName: string, newValue: string ) => void;
}

function CountrySpecificPaymentField< P >( {
	fieldName,
	allowedFieldNames,
	componentClass,
	props,
	disabled,
	getErrorMessages,
	getFieldValue,
	handleFieldChange,
}: CountrySpecificPaymentFieldProps< P > ) {
	if ( ! allowedFieldNames.includes( fieldName ) ) {
		return null;
	}
	const errorMessages = getErrorMessages( fieldName );
	return createField( {
		fieldName,
		componentClass,
		props,
		value: getFieldValue( fieldName ),
		errorMessage: errorMessages.length > 0 ? errorMessages[ 0 ] : '',
		handleFieldChange,
		disabled: disabled ?? false,
	} );
}

function CountrySpecificPaymentFieldsUnstyled( {
	countriesList,
	countryCode,
	disableFields,
	getErrorMessages,
	getFieldValue,
	handleFieldChange,
}: CountrySpecificPaymentFieldsProps ) {
	const [ phoneCountryCode, setPhoneCountryCode ] = useState( '' );
	const fieldNames = getFieldNamesForCountry( countryCode );
	const translate = useTranslate();

	const handlePhoneFieldChange = ( {
		value,
		countryCode: newPhoneCountryCode,
	}: {
		value: string;
		countryCode: string;
	} ) => {
		handleFieldChange( 'phone-number', value );
		setPhoneCountryCode( newPhoneCountryCode );
	};

	const getPanNumberPopover = () => {
		const popoverText = translate( 'Paying with %(indiaPaymentMethods)s requires a PAN number.', {
			comment: 'indiaPaymentMethods are local payment methods in India.',
			args: {
				indiaPaymentMethods: paymentMethodName( 'netbanking' ),
			},
		} );
		return (
			<InfoPopover position="right" className="checkout__pan-number-popover">
				{ popoverText }
			</InfoPopover>
		);
	};

	const countryData = countriesList.find(
		( countryElement ) => countryElement.code === countryCode
	);
	const countryName = countryData && countryData.name ? countryData.name : '';
	const containerClassName = classNames(
		'checkout__country-payment-fields',
		`checkout__country-${ countryCode.toLowerCase() }`
	);

	return (
		<div className={ containerClassName }>
			<span key="country-required-fields" className="checkout__form-info-text">
				{ countryName &&
					translate( 'The following fields are also required for payments in %(countryName)s', {
						args: {
							countryName,
						},
					} ) }
			</span>

			<CountrySpecificPaymentField
				fieldName="document"
				allowedFieldNames={ fieldNames }
				componentClass={ Input }
				getFieldValue={ getFieldValue }
				getErrorMessages={ getErrorMessages }
				handleFieldChange={ handleFieldChange }
				disabled={ disableFields }
				props={ {
					label: translate( 'Taxpayer Identification Number', {
						comment:
							'Individual taxpayer registry identification required ' +
							'for Brazilian payment methods on credit card form',
					} ),
				} }
			/>

			<CountrySpecificPaymentField
				fieldName="nik"
				allowedFieldNames={ fieldNames }
				componentClass={ Input }
				getFieldValue={ getFieldValue }
				getErrorMessages={ getErrorMessages }
				handleFieldChange={ handleFieldChange }
				disabled={ disableFields }
				props={ {
					label: translate( 'NIK - Indonesia Identity Card Number', {
						comment:
							'NIK - Indonesia Identity Card Number required for Indonesian payment methods.',
					} ),
				} }
			/>

			<CountrySpecificPaymentField
				fieldName="pan"
				allowedFieldNames={ fieldNames }
				componentClass={ Input }
				getFieldValue={ getFieldValue }
				getErrorMessages={ getErrorMessages }
				handleFieldChange={ handleFieldChange }
				disabled={ disableFields }
				props={ {
					placeholder: ' ',
					label: translate( 'PAN Number {{panNumberPopover/}}', {
						comment: 'India PAN number ',
						components: {
							panNumberPopover: getPanNumberPopover(),
						},
					} ),
				} }
			/>

			<CountrySpecificPaymentField
				fieldName="gstin"
				allowedFieldNames={ fieldNames }
				componentClass={ Input }
				getFieldValue={ getFieldValue }
				getErrorMessages={ getErrorMessages }
				handleFieldChange={ handleFieldChange }
				disabled={ disableFields }
				props={ {
					placeholder: ' ',
					label: translate( 'GSTIN (optional)', {
						comment: 'India PAN number ',
					} ),
				} }
			/>

			<CountrySpecificPaymentField
				fieldName="phone-number"
				allowedFieldNames={ fieldNames }
				componentClass={ FormPhoneMediaInput }
				getFieldValue={ getFieldValue }
				getErrorMessages={ getErrorMessages }
				handleFieldChange={ handleFieldChange }
				disabled={ disableFields }
				props={ {
					// This is the same as what would be automatically provided by
					// CountrySpecificPaymentField but is required here by TypeScript to
					// validate the component props.
					value: getFieldValue( 'phone-number' ),
					// This will override the handleFieldChange prop.
					onChange: handlePhoneFieldChange,
					countriesList,
					// If the user has manually selected a country for the phone
					// number, use that, but otherwise default this to the same
					// country as the billing address.
					countryCode: phoneCountryCode || countryCode,
					label: translate( 'Phone' ),
				} }
			/>

			<CountrySpecificPaymentField
				fieldName="address-1"
				allowedFieldNames={ fieldNames }
				componentClass={ Input }
				getFieldValue={ getFieldValue }
				getErrorMessages={ getErrorMessages }
				handleFieldChange={ handleFieldChange }
				disabled={ disableFields }
				props={ {
					maxLength: 40,
					label: translate( 'Address' ),
				} }
			/>

			<CountrySpecificPaymentField
				fieldName="street-number"
				allowedFieldNames={ fieldNames }
				componentClass={ Input }
				getFieldValue={ getFieldValue }
				getErrorMessages={ getErrorMessages }
				handleFieldChange={ handleFieldChange }
				disabled={ disableFields }
				props={ {
					inputMode: 'numeric',
					label: translate( 'Street Number', {
						comment: 'Street number associated with address on credit card form',
					} ),
				} }
			/>

			<CountrySpecificPaymentField
				fieldName="address-2"
				allowedFieldNames={ fieldNames }
				componentClass={ HiddenInput }
				getFieldValue={ getFieldValue }
				getErrorMessages={ getErrorMessages }
				handleFieldChange={ handleFieldChange }
				disabled={ disableFields }
				props={ {
					maxLength: 40,
					label: translate( 'Address Line 2' ),
					text: translate( '+ Add Address Line 2' ),
				} }
			/>

			<CountrySpecificPaymentField
				fieldName="city"
				allowedFieldNames={ fieldNames }
				componentClass={ Input }
				getFieldValue={ getFieldValue }
				getErrorMessages={ getErrorMessages }
				handleFieldChange={ handleFieldChange }
				disabled={ disableFields }
				props={ {
					label: translate( 'City' ),
				} }
			/>

			<div className="checkout__form-state-field">
				<CountrySpecificPaymentField
					fieldName="state"
					allowedFieldNames={ fieldNames }
					componentClass={ StateSelect }
					getFieldValue={ getFieldValue }
					getErrorMessages={ getErrorMessages }
					handleFieldChange={ handleFieldChange }
					disabled={ disableFields }
					props={ {
						countryCode,
						label: translate( 'State' ),
					} }
				/>
			</div>

			<CountrySpecificPaymentField
				fieldName="postal-code"
				allowedFieldNames={ fieldNames }
				componentClass={ Input }
				getFieldValue={ getFieldValue }
				getErrorMessages={ getErrorMessages }
				handleFieldChange={ handleFieldChange }
				disabled={ disableFields }
				props={ {
					label: translate( 'Postal Code' ),
				} }
			/>
		</div>
	);
}
