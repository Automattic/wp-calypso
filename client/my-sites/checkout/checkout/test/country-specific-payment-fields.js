/**
 * @format
 * @jest-environment jsdom
 */
/**
 * External dependencies
 */
import React from 'react';
import { shallow } from 'enzyme';
import { identity } from 'lodash';

/**
 * Internal dependencies
 */

import { CountrySpecificPaymentFields } from '../country-specific-payment-fields';

// Gets rid of warnings such as 'UnhandledPromiseRejectionWarning: Error: No available storage method found.'
jest.mock( 'lib/user', () => () => {} );

const defaultProps = {
	countryCode: 'BR',
	countriesList: [
		{
			code: 'BR',
			name: 'Brazil',
		},
	],
	getErrorMessage: jest.fn(),
	getFieldValue: jest.fn(),
	handleFieldChange: jest.fn(),
	fieldClassName: 'country-brazil',
	translate: identity,
};

describe( '<CountrySpecificPaymentFields />', () => {
	test( 'should render', () => {
		const wrapper = shallow( <CountrySpecificPaymentFields { ...defaultProps } /> );
		expect( wrapper ).toMatchSnapshot();
	} );

	test( 'should call this.props.handleFieldChange when updating field', () => {
		const wrapper = shallow( <CountrySpecificPaymentFields { ...defaultProps } /> );
		const documentInput = wrapper.find( '[name="document"]' );
		const event = { target: { name: 'document', value: 'spam' } };
		documentInput.simulate( 'change', event );
		expect( defaultProps.handleFieldChange ).toBeCalledWith( 'document', 'spam' );
	} );

	test( 'should disable fields', () => {
		const wrapper = shallow( <CountrySpecificPaymentFields { ...defaultProps } /> );
		expect(
			wrapper
				.find( 'Input' )
				.first()
				.props().disabled
		).toEqual( false );
		wrapper.setProps( { disableFields: true } );
		expect(
			wrapper
				.find( 'Input' )
				.first()
				.props().disabled
		).toEqual( true );
	} );
} );
