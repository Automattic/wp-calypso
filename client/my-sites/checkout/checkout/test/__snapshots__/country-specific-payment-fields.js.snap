// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CountrySpecificPaymentFields /> should render 1`] = `
<div
  className="checkout__country-payment-fields checkout__country-br"
>
  <span
    className="checkout__form-info-text"
    key="country-required-fields"
  >
    The following fields are also required for payments in %(countryName)s
  </span>
  <Input
    additionalClasses="checkout__checkout-field country-brazil"
    autoComplete="off"
    autoFocus={false}
    disabled={false}
    isError={false}
    label="Taxpayer Identification Number"
    labelClass="checkout__form-label"
    name="document"
    onBlur={[Function]}
    onChange={[Function]}
    value=""
  />
  <FormPhoneMediaInput
    additionalClasses="checkout__checkout-field country-brazil"
    autoComplete="off"
    countriesList={
      Array [
        Object {
          "code": "BR",
          "name": "Brazil",
        },
      ]
    }
    countryCode="BR"
    disabled={false}
    isError={false}
    label="Phone"
    labelClass="checkout__form-label"
    name="phone-number"
    onBlur={[Function]}
    onChange={[Function]}
    value=""
  />
  <Input
    additionalClasses="checkout__checkout-field country-brazil"
    autoComplete="off"
    autoFocus={false}
    disabled={false}
    isError={false}
    label="Address"
    labelClass="checkout__form-label"
    maxLength={40}
    name="address-1"
    onBlur={[Function]}
    onChange={[Function]}
    value=""
  />
  <Input
    additionalClasses="checkout__checkout-field country-brazil"
    autoComplete="off"
    autoFocus={false}
    disabled={false}
    inputMode="numeric"
    isError={false}
    label="Street Number"
    labelClass="checkout__form-label"
    name="street-number"
    onBlur={[Function]}
    onChange={[Function]}
    value=""
  />
  <HiddenInput
    additionalClasses="checkout__checkout-field country-brazil"
    autoComplete="off"
    disabled={false}
    isError={false}
    label="Address Line 2"
    labelClass="checkout__form-label"
    maxLength={40}
    name="address-2"
    onBlur={[Function]}
    onChange={[Function]}
    text="+ Add Address Line 2"
    value=""
  />
  <Input
    additionalClasses="checkout__checkout-field country-brazil"
    autoComplete="off"
    autoFocus={false}
    disabled={false}
    isError={false}
    label="City"
    labelClass="checkout__form-label"
    name="city"
    onBlur={[Function]}
    onChange={[Function]}
    value=""
  />
  <div
    className="checkout__form-state-field"
  >
    <Connect(Localized(StateSelect))
      additionalClasses="checkout__checkout-field country-brazil"
      autoComplete="off"
      countryCode="BR"
      disabled={false}
      isError={false}
      label="State"
      labelClass="checkout__form-label"
      name="state"
      onBlur={[Function]}
      onChange={[Function]}
      value=""
    />
  </div>
  <Input
    additionalClasses="checkout__checkout-field country-brazil"
    autoComplete="off"
    autoFocus={false}
    disabled={false}
    isError={false}
    label="Postal Code"
    labelClass="checkout__form-label"
    name="postal-code"
    onBlur={[Function]}
    onChange={[Function]}
    value=""
  />
</div>
`;
